File: packages/sdk/ios/core/BLEIntSDK/Classes/Protocol/ProtocolManager.swift
//
//  BLEProtocolManager
//
//  Copyright © 2022 2hire.io. All rights reserved.
//

import CryptoKit
import Foundation
import K1
import Logging
import os.log

public struct ProtocolCommandResponse {
    internal init(success: Bool, additionalPayload: [UInt8]) {
        self.success = success
        self.additionalPayload = additionalPayload
    }

    public let success: Bool
    public let additionalPayload: [UInt8]

    internal var description: String {
        "Command was \(self.success ? "successful" : "unsuccessful") with additionalPayload \(Data(self.additionalPayload).base64EncodedString())"
    }
}

internal typealias ProtocolResult = Result<ProtocolCommandResponse, ProtocolErrorCode>

internal class ProtocolManager {
    private var privateKey: PrivateKey!
    private var publicKey: PublicKey!

    private var writable: WritableTL!
    private(set) var writableState: WritableTLState?

    private var writeBuffer: [[UInt8]] = []
    private var readBuffer: [UInt8] = []

    private var writeContinuation: CheckedContinuation<ProtocolResult, Error>?
    private var connectionContinuation: CheckedContinuation<Void, Error>?

    private static let logger = LoggingUtil.logger

    init(with writable: WritableTL, privateKey: PrivateKey, publicKey: PublicKey) {
        self.privateKey = privateKey
        self.publicKey = publicKey

        self.writable = writable
        self.writable.writableDelegate = self
        self.writable.connectableDelegate = self
    }

    func setPrivateKey(_ key: PrivateKey) {
        self.privateKey = key
    }

    func connect(to macAddress: String) async throws {
        return try await withCheckedThrowingContinuation { cont in
            guard self.connectionContinuation == nil else {
                cont.resume(throwing: ProtocolError.ApiMisuse)
                return
            }

            do {
                self.connectionContinuation = cont
                try self.writable.connect(to: macAddress)
            }
            catch {
                cont.resume(throwing: error)
                self.connectionContinuation = nil
            }
        }
    }

    func startSession(withAccess data: [UInt8]) async throws -> ProtocolResult {
        return try await self.withThrowingWriteContinuation {
            let personalPublicKey = try self.privateKey.publicKey.rawRepresentation(format: .compressed)

            self.writeBuffer = [
                ProtocolFrame.SessionStart, personalPublicKey + data, ProtocolFrame.SessionEnd,
            ]

            Self.logger.debug(
                "Start session data: \(Data(writeBuffer.joined()).hexEncodedString)",
                metadata: .protocol
            )
            try self.write()
        }
    }

    func sendCommand(withPayload data: [UInt8]) async throws -> ProtocolResult {
        return try await self.withThrowingWriteContinuation {
            do {
                let commandPayload = CommandRequestPayload(data: data)

                let encryptedPacket = try EncryptedCommandPacket.encrypt(
                    commandPayload.encode(),
                    version: ProtocolVersion,
                    privateKey: privateKey,
                    publicKey: publicKey
                ).encode()

                self.writeBuffer = [
                    ProtocolFrame.CommandStart, encryptedPacket, ProtocolFrame.CommandEnd,
                ]
            }
            catch {
                Self.logger.error(
                    "Error while encrypting command data: \(error.localizedDescription)",
                    metadata: .protocol
                )
                throw ProtocolError.Crypto
            }

            Self.logger.info(
                "Sending command data: \(Data(writeBuffer.joined()).hexEncodedString)",
                metadata: .protocol
            )

            try self.write()
        }
    }

    private func withThrowingWriteContinuation(_ body: () throws -> Void) async throws -> ProtocolResult {
        return try await withCheckedThrowingContinuation { cont in
            guard self.writeContinuation == nil else {
                cont.resume(throwing: ProtocolError.ApiMisuse)
                return
            }

            do {
                self.writeContinuation = cont

                try body()
            }
            catch {
                cont.resume(throwing: error)
                self.writeContinuation = nil
            }
        }
    }
}

// MARK: Reading and writing handlers

extension ProtocolManager: WritableTLDelegate {
    private func write() throws {
        guard let nextChunk = self.writeBuffer.first else {
            Self.logger.info("Chunk is empty, starting read", metadata: .protocol)

            do {
                try self.writable.startReading()
            }
            catch {
                Self.logger.error("Error while writing", metadata: .protocol)

                self.writeContinuation?.resume(throwing: error)
                self.writeContinuation = nil
            }
            return
        }

        guard self.writableState == .Connected else {
            Self.logger.info(
                "Writable is not in connected state: \((self.writableState ?? .Reading).description)",
                metadata: .protocol

            )

            self.writeContinuation?.resume(throwing: ProtocolError.Writable)
            self.writeContinuation = nil

            return
        }

        Self.logger.debug("Writing data: \(Data(nextChunk).hexEncodedString)", metadata: .protocol)

        try self.writable.write(data: nextChunk)
    }

    func writable(didWrite remainingData: [UInt8], _ error: Error?) {
        if let error = error {
            Self.logger.error("Error while writing \(error.localizedDescription)", metadata: .protocol)

            self.writeContinuation?.resume(throwing: error)
            self.writeContinuation = nil

            return
        }

        guard remainingData.isEmpty else {
            Self.logger.error(
                "Error while writing, data is not empty: \(remainingData.description)",
                metadata: .protocol
            )

            self.writeContinuation?.resume(throwing: ProtocolError.Generic)
            self.writeContinuation = nil

            return
        }

        self.writeBuffer = [[UInt8]](self.writeBuffer.dropFirst())

        do {
            try self.write()
        }
        catch {
            Self.logger.error("Error while writing \(remainingData.description)", metadata: .protocol)

            self.writeContinuation?.resume(throwing: error)
            self.writeContinuation = nil
        }
    }

    func writable(didReceive data: [UInt8]?, _ error: Error?) {
        guard let receivedData = data, error == nil else {
            Self.logger.error("Error while reading \(error.debugDescription)", metadata: .protocol)

            self.writeContinuation?.resume(throwing: error ?? ProtocolError.Generic)
            self.writeContinuation = nil

            return
        }

        Self.logger.debug("Did receive \(Data(receivedData).hexEncodedString)", metadata: .protocol)

        switch receivedData {
        case ProtocolFrame.SessionStart, ProtocolFrame.CommandStart:
            self.readBuffer = [UInt8](receivedData)
        case ProtocolFrame.SessionEnd, ProtocolFrame.CommandEnd:
            self.readBuffer += receivedData

            Self.logger.debug(
                "Read end frame, closing",
                metadata: .protocol
            )

            do {
                try self.writable.stopReading()
            }
            catch {
                Self.logger.error(
                    "Error while stop reading \(error.localizedDescription)",
                    metadata: .protocol
                )

                self.writeContinuation?.resume(throwing: error)
                self.writeContinuation = nil
            }
        default:
            self.readBuffer += receivedData
        }
    }

    func writable(didStopReading: Bool, _ error: Error?) {
        do {
            Self.logger.info("Received data: \(Data(self.readBuffer).hexEncodedString)", metadata: .protocol)

            let startFrame = [UInt8](self.readBuffer.prefix(4))
            let payload: [UInt8] = self.readBuffer.dropFirst(4).dropLast(4)

            Self.logger.debug(
                "DidStopReading start frame: \(startFrame.description)",
                metadata: .protocol
            )

            switch startFrame {
            case ProtocolFrame.SessionStart, ProtocolFrame.CommandStart:
                self.processEncryptedCommandResponse(try EncryptedCommandPacket(from: payload))
            default:
                throw ProtocolError.InvalidData("Invalid protocol start frame")
            }
        }
        catch {
            Self.logger.error(
                "Error while parsing packets \(error.localizedDescription)",
                metadata: .protocol
            )
            self.writeContinuation?.resume(throwing: error)
            self.writeContinuation = nil
        }
    }

    func connection(didChangeState state: WritableTLState) {
        Self.logger.info(
            "Did change status: \(self.writableState?.description ?? "nil") -> \(state.description)",
            metadata: .protocol
        )
        self.writableState = state
    }
}

// MARK: Connection handlers

extension ProtocolManager: ConnectableTLDelegate {
    func create(didCreate data: Any?, _ error: Error?) {
        guard error == nil else {
            Self.logger.error(
                "Creation Error \(error?.localizedDescription ?? "Error not set")",
                metadata: .protocol
            )

            self.connectionContinuation?.resume(throwing: ProtocolError.Writable)
            self.connectionContinuation = nil

            return
        }

        Self.logger.info("Writable created", metadata: .protocol)
    }

    func connect(didConnect data: Any?, _ error: Error?) {
        guard error == nil else {
            Self.logger.error(
                "Connection Error \(error?.localizedDescription ?? "Error not set")",
                metadata: .protocol
            )

            self.connectionContinuation?.resume(throwing: error ?? ProtocolError.Writable)
            self.connectionContinuation = nil

            return
        }

        Self.logger.info("Writable connected", metadata: .protocol)

        self.connectionContinuation?.resume()
        self.connectionContinuation = nil
    }
}

// MARK: Packet handlers

extension ProtocolManager {
    private func processEncryptedCommandResponse(_ packet: EncryptedCommandPacket) {
        Self.logger.debug(
            "Received encrypted command response \(Data(packet.data).hexEncodedString)",
            metadata: .protocol
        )

        do {
            let decryptedData = try packet.decrypt(privateKey: self.privateKey, publicKey: self.publicKey)
            Self.logger.debug(
                "Decrypted command data: \(Data(decryptedData).hexEncodedString)",
                metadata: .protocol
            )

            let commandPayload = try CommandResponsePayload.init(from: decryptedData)

            if commandPayload.commandIdentifier != .Error {
                self.writeContinuation?.resume(
                    returning: .success(
                        .init(success: commandPayload.commandIdentifier == .Ack, additionalPayload: commandPayload.data)
                    )
                )
            }
            else {
                let errorPayload = try ErrorCommandPayload.init(from: commandPayload.data)

                self.writeContinuation?.resume(returning: .failure(errorPayload.errorCode))
            }
        }
        catch let error as CryptoKitError {
            Self.logger.error("CryptoKit error \(error.description)", metadata: .protocol)
            self.writeContinuation?.resume(throwing: error)
        }
        catch {
            Self.logger.error("Error while decrypting data \(error.localizedDescription)", metadata: .protocol)
            self.writeContinuation?.resume(throwing: error)
        }

        self.writeContinuation = nil
    }
}

// MARK: extensions
extension ProtocolErrorCode: Error {}

extension Logging.Logger.Metadata {
    fileprivate static var `protocol`: Self {
        var metadata: Self = ["category": "🔀 ProtocolManager"]

        if let requestId = Self.requestId {
            metadata["requestId"] = "\(requestId)"
        }

        return metadata
    }
}

File: packages/sdk/android/core/src/main/java/io/twohire/bleintsdk/protocol/ProtocolManager.kt
package io.twohire.bleintsdk.protocol

import android.content.*
import android.os.IBinder
import android.util.Log
import androidx.core.content.ContextCompat
import io.twohire.bleintsdk.bluetooth.BluetoothAction
import io.twohire.bleintsdk.bluetooth.BluetoothError
import io.twohire.bleintsdk.bluetooth.BluetoothLeService
import io.twohire.bleintsdk.crypto.CryptoHelper
import io.twohire.bleintsdk.crypto.ECKeyPair
import io.twohire.bleintsdk.utils.toHex
import org.spongycastle.jce.interfaces.ECPublicKey
import kotlin.coroutines.Continuation
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import kotlin.coroutines.suspendCoroutine

internal class ProtocolManager private constructor(
    private var keyPair: ECKeyPair,
    private var publicKey: ECPublicKey
) {
    private val tag = "${ProtocolManager::class.simpleName}@${System.identityHashCode(this)}"

    private var bluetoothLeService: BluetoothLeService? = null
    private var bluetoothLeServiceConn: BluetoothLeServiceConnection? = null
    var writableState: WritableTLState? = null

    fun setKeyPair(keyPair: ECKeyPair) {
        this.keyPair = keyPair
    }

    protected fun setPublicKey(keyPair: ECPublicKey) {
        this.publicKey = keyPair
    }

    private inner class BluetoothLeServiceConnection : ServiceConnection {
        override fun onServiceConnected(componentName: ComponentName?, binder: IBinder?) {
            Log.d(tag, "ServiceConnection: connected to $componentName")

            val service = (binder as BluetoothLeService.LocalBinder).service

            service.setDataMerger(ProtocolDataMerger())
            this@ProtocolManager.bluetoothLeService = service
        }

        override fun onServiceDisconnected(componentName: ComponentName?) {
            Log.w(tag, "ServiceConnection: disconnected from $componentName")

            this@ProtocolManager.bluetoothLeService = null
        }

        override fun onBindingDied(componentName: ComponentName?) {
            Log.w(tag, "ServiceConnection: onBindingDied $componentName")
        }

        override fun onNullBinding(componentName: ComponentName?) {
            Log.w(tag, "ServiceConnection: onNullBinding $componentName")
        }
    }

    private val broadcastReceiver = object : BroadcastReceiver() {
        override fun onReceive(p0: Context?, intent: Intent?) {
            Log.d(tag, "[Action]: BluetoothAction.${intent?.action}")

            when (intent?.action) {
                BluetoothAction.ACTION_STATE_CHANGED -> {
                    intent.getStringExtra(BluetoothAction.ACTION_EXTRA_DATA)?.let {
                        WritableTLState.fromName(it)?.let { state ->
                            Log.d(
                                tag,
                                "Did change state: ${this@ProtocolManager.writableState?.name} -> ${state.name}"
                            )
                            this@ProtocolManager.writableState = state
                        }
                    }
                }

                BluetoothAction.ACTION_CONNECT_COMPLETE -> {
                    Log.d(tag, "Writable connected")
                    this@ProtocolManager.connectionContinuation?.resume(true)
                        .also { this@ProtocolManager.connectionContinuation = null }
                }

                BluetoothAction.ACTION_CONNECT_ERROR -> {
                    Log.e(tag, "Connection error")

                    this@ProtocolManager.connectionContinuation?.resumeWithException(
                        IllegalStateException(
                            "${ProtocolError.WRITABLE.name}: (${
                                intent.getIntExtra(
                                    BluetoothAction.ACTION_EXTRA_DATA,
                                    -1
                                )
                            })"
                        )
                    )
                        .also {
                            this@ProtocolManager.connectionContinuation = null
                        }
                }

                BluetoothAction.ACTION_CONNECT_TIMEOUT -> {
                    Log.e(tag, "Connection timeout error")

                    this@ProtocolManager.connectionContinuation?.resumeWithException(
                        IllegalStateException(
                            BluetoothError.TIMEOUT_ERROR.name
                        )
                    )
                        .also {
                            this@ProtocolManager.connectionContinuation = null
                        }
                }

                BluetoothAction.ACTION_WRITE_COMPLETE -> {
                    try {
                        this@ProtocolManager.writeBuffer =
                            this@ProtocolManager.writeBuffer.drop(1).toTypedArray()
                        this@ProtocolManager.write()
                    } catch (error: Exception) {
                        Log.e(tag, "Error while writing ${error.message}")

                        this@ProtocolManager.writeContinuation?.resumeWithException(error)
                            .also { this@ProtocolManager.writeContinuation = null }
                    }
                }

                BluetoothAction.ACTION_WRITE_ERROR -> {
                    Log.e(tag, "Error while writing")

                    this@ProtocolManager.writeContinuation?.resumeWithException(
                        IllegalStateException(
                            "${ProtocolError.WRITABLE.name}: (${
                                intent.getIntExtra(
                                    BluetoothAction.ACTION_EXTRA_DATA,
                                    -1
                                )
                            })"
                        )
                    ).also { this@ProtocolManager.writeContinuation = null }
                }

                BluetoothAction.ACTION_READ_DATA -> {
                    val data = intent.getByteArrayExtra(BluetoothAction.ACTION_EXTRA_DATA)

                    if (data != null) {
                        Log.d(tag, "Incoming data: ${data.contentToString()}, closing")

                        try {
                            this@ProtocolManager.readBuffer = data
                            this@ProtocolManager.bluetoothLeService?.stopReading()
                        } catch (error: Exception) {
                            Log.e(tag, "Error while stop reading ${error.message}")

                            this@ProtocolManager.writeContinuation?.resumeWithException(
                                error
                            ).also { this@ProtocolManager.writeContinuation = null }
                        }
                    } else {
                        Log.e(tag, "Error while reading")

                        this@ProtocolManager.writeContinuation?.resumeWithException(
                            IllegalStateException(ProtocolError.WRITABLE.name)
                        ).also { this@ProtocolManager.writeContinuation = null }
                    }
                }

                BluetoothAction.ACTION_READ_ERROR -> {
                    Log.e(tag, "Error while reading")

                    this@ProtocolManager.writeContinuation?.resumeWithException(
                        IllegalStateException(
                            "${ProtocolError.WRITABLE.name}: (${
                                intent.getIntExtra(
                                    BluetoothAction.ACTION_EXTRA_DATA,
                                    -1
                                )
                            })"
                        )
                    )
                        .also { this@ProtocolManager.writeContinuation = null }
                }

                BluetoothAction.ACTION_READ_COMPLETE -> {
                    try {
                        Log.d(tag, "Stop read data: ${readBuffer.toHex()}")

                        val payload = arrayOf(
                            ProtocolFrame.SESSION_START.rawValue,
                            ProtocolFrame.COMMAND_START.rawValue
                        ).firstNotNullOf {
                            val firstMatch = readBuffer.slice(it.indices).toByteArray()
                            val secondMatch =
                                readBuffer.slice(it.indices.last + 1 until it.indices.last + 1 + it.size)
                                    .toByteArray()

                            if (secondMatch.contentEquals(it)) {
                                return@firstNotNullOf readBuffer.drop(it.size * 2).toByteArray()
                            } else if (firstMatch.contentEquals(it)) {
                                return@firstNotNullOf readBuffer.drop(it.size).toByteArray()
                            }

                            null
                        }.dropLast(4).toByteArray()

                        this@ProtocolManager.processCommandResponse(
                            EncryptedCommandPacket.create(
                                payload
                            )
                        )

                    } catch (error: NoSuchElementException) {
                        Log.e(tag, "Start frame not found")

                        this@ProtocolManager.writeContinuation?.resumeWithException(
                            IllegalStateException(ProtocolError.INVALID_DATA.name)
                        )
                            .also { this@ProtocolManager.writeContinuation = null }
                    } catch (error: Exception) {
                        Log.e(tag, "Error while stop reading (${error.message})")

                        this@ProtocolManager.writeContinuation?.resumeWithException(error)
                            .also { this@ProtocolManager.writeContinuation = null }
                    }
                }
            }
        }
    }

    private var connectionContinuation: Continuation<Boolean>? = null
    private var writeContinuation: Continuation<Result<ProtocolResponse>>? = null

    private var writeBuffer = arrayOf(ByteArray(0))
    private var readBuffer = ByteArray(0)

    private constructor(
        context: Context,
        keyPair: ECKeyPair,
        publicKey: ECPublicKey
    ) : this(keyPair, publicKey) {

        val connection = BluetoothLeServiceConnection()
        val status =
            context.bindService(
                Intent(
                    BluetoothAction.ACTION_BLE_SERVICE,
                    null,
                    context,
                    BluetoothLeService::class.java
                ), connection, Context.BIND_AUTO_CREATE
            )

        Log.d(tag, "ServiceStatus: $status")

        if (!status) {
            throw IllegalStateException(ProtocolError.INTERNAL.name)
        } else {
            this.bluetoothLeServiceConn = connection
        }

        ContextCompat.registerReceiver(
            context,
            this.broadcastReceiver,
            IntentFilter().apply {
                addAction(BluetoothAction.ACTION_STATE_CHANGED)
                addAction(BluetoothAction.ACTION_CONNECT_COMPLETE)
                addAction(BluetoothAction.ACTION_CONNECT_TIMEOUT)
                addAction(BluetoothAction.ACTION_CONNECT_ERROR)
                addAction(BluetoothAction.ACTION_WRITE_COMPLETE)
                addAction(BluetoothAction.ACTION_WRITE_ERROR)
                addAction(BluetoothAction.ACTION_READ_COMPLETE)
                addAction(BluetoothAction.ACTION_READ_DATA)
                addAction(BluetoothAction.ACTION_READ_ERROR)
            },
            ContextCompat.RECEIVER_EXPORTED
        )
    }

    suspend fun connect(id: String) = suspendCoroutine<Boolean> { cont ->
        if (this@ProtocolManager.connectionContinuation !== null) {
            cont.resumeWithException(IllegalStateException(ProtocolError.API_MISUSE.name))
        } else {
            try {
                this@ProtocolManager.connectionContinuation = cont
                this@ProtocolManager.bluetoothLeService?.connect(id)
            } catch (error: Exception) {
                cont.resumeWithException(error)
                this@ProtocolManager.connectionContinuation = null
            }
        }
    }

    suspend fun startSession(accessData: ByteArray) = withThrowingWriteContinuation { ->
        val personalPublicKey =
            CryptoHelper.compactPublicKey(this@ProtocolManager.keyPair.publicKey)

        this@ProtocolManager.writeBuffer = arrayOf(
            ProtocolFrame.SESSION_START.rawValue,
            personalPublicKey + accessData,
            ProtocolFrame.SESSION_END.rawValue,
        )

        Log.d(
            tag,
            "Start session data ${
                writeBuffer.map { it.contentToString() }.toTypedArray().contentToString()
            }"
        )

        this@ProtocolManager.write()
    }

    suspend fun sendCommand(payload: ByteArray) = withThrowingWriteContinuation {
        try {
            val commandPayload = CommandRequestPayload(payload).encode()

            val encryptedPacket = EncryptedCommandPacket.encrypt(
                commandPayload,
                PROTOCOL_VERSION,
                keyPair.privateKey,
                publicKey
            ).encode()

            this@ProtocolManager.writeBuffer = arrayOf(
                ProtocolFrame.COMMAND_START.rawValue,
                encryptedPacket,
                ProtocolFrame.COMMAND_END.rawValue
            )
        } catch (exception: Exception) {
            Log.e(tag, "Error while encrypting command data: ${exception.message}")
            throw IllegalStateException(ProtocolError.CRYPTO.name)
        }

        Log.d(
            tag,
            "Sending command data: ${
                writeBuffer.map { it.contentToString() }.toTypedArray().contentToString()
            }"
        )

        this@ProtocolManager.write()
    }

    private fun processCommandResponse(packet: EncryptedCommandPacket) =
        this@ProtocolManager.writeContinuation?.let {
            Log.d(
                tag,
                "Received encrypted command response: ${packet.data.toHex()}"
            )

            val decryptedData = packet.decrypt(
                this@ProtocolManager.keyPair.privateKey,
                this@ProtocolManager.publicKey
            )
            Log.d(
                tag,
                "Decrypted command data: ${decryptedData.toHex()}"
            )

            val commandPayload = CommandResponsePayload.create(decryptedData)

            if (commandPayload.commandIdentifier != CommandIdentifier.ERROR) {
                it.resume(
                    Result.success(
                        ProtocolResponse(
                            commandPayload.commandIdentifier == CommandIdentifier.ACK,
                            commandPayload.data
                        )
                    )
                )
            } else {
                val errorPayload = ErrorCommandPayload.create(commandPayload.data)
                it.resume(Result.failure(ProtocolErrorCodeException(errorPayload.errorCode)))
            }
        }.also {
            this@ProtocolManager.writeContinuation = null
        }

    private fun write() {
        try {
            val service =
                this.bluetoothLeService ?: throw IllegalStateException(ProtocolError.WRITABLE.name)

            if (this.writeBuffer.isEmpty()) {
                Log.d(tag, "Buffer is empty, starting read")
                service.startReading()
            } else if (this@ProtocolManager.writableState != WritableTLState.Connected) {
                Log.e(tag, "Writable is not in connected state")

                throw IllegalStateException(ProtocolError.WRITABLE.name)
            } else {
                service.write(writeBuffer.first())
            }
        } catch (error: Exception) {
            Log.e(tag, "Error while writing")

            this.writeContinuation?.resumeWithException(error).also {
                this.writeContinuation = null
            }
        }
    }

    private suspend fun withThrowingWriteContinuation(body: () -> Unit) =
        suspendCoroutine<Result<ProtocolResponse>> { cont ->
            if (this@ProtocolManager.writeContinuation !== null) {
                cont.resumeWithException(IllegalStateException(ProtocolError.API_MISUSE.name))
            }

            try {
                this@ProtocolManager.writeContinuation = cont

                body()
            } catch (exception: Exception) {
                cont.resumeWithException(exception)
                this@ProtocolManager.writeContinuation = null
            }
        }

    companion object {
        private val TAG = "${ProtocolManager::class.simpleName}"

        private var instance: ProtocolManager? = null

        fun getInstance(
            context: Context,
            keyPair: ECKeyPair,
            publicKey: ECPublicKey
        ): ProtocolManager {
            var instance = this.instance

            if (instance == null) {
                instance = ProtocolManager(context, keyPair, publicKey)
                this.instance = instance
            } else {
                instance.setKeyPair(keyPair)
                instance.setPublicKey(publicKey)
            }

            return instance
        }
    }
}

internal class ProtocolErrorCodeException(val errorCode: ProtocolErrorCode) : Throwable()